TypeScript

thompsongl opened the pull request:
To do:
* Come up with a better name than `ColorModeSwitch`?
* ~Make `EuiComputedTheme` work again~
* ~Discuss how this affects custom color modes~ See https://github.com/elastic/eui/pull/4643#issuecomment-811164296

~There are a few colors missing from the computed theme type. It's definitely fixable but depends on what shape the colors section takes.~ Fixed with _more_ TypeScript

thompsongl committed:
wip ts for theme building

thompsongl committed:
better computed

thompsongl committed:
computed type from uncomputed type; ams types

thompsongl commented:
This is ready to look at. The Amsterdam theme is still incomplete, but the types are much better and it'll run error-free

cchaos commented on src/global_styling/variables/_colors.ts:
```diff
@@ -26,12 +27,9 @@ import {
-  textSubdued: any;
+  text: ColorModeSwitch;
+  title: ColorModeSwitch;
+  textSubdued: ColorModeSwitch;
   textPrimary: string;
```
So is `ColorModeSwitch` flexible enough where the value could either be the `LIGHT/DARK` keys OR a simple `string`? Like I could also use it here:
```suggestion
  textSubdued: ColorModeSwitch;
  textPrimary: ColorModeSwitch;
```

Even though the default theme is setting it as a single `string`?

thompsongl replied:
You can definitely do `ColorModeSwitch | string`.
Do you think that this will be a common use case? It might be possible to automatically apply the union for primitive types like `string` and `number`.

cchaos replied:
Ah yes, we can certainly do `ColorModeSwitch | string`. But does `ColorModeSwitch` then also strongly type that the value of the LIGHT/DARK keys **also** have to be strings? Maybe that's what you're getting at with the primitive thing?

But yes, I think it'll be most flexible to use this `ColorModeSwitch` for just about every value that is most likely a color . Even if our own themes might use a single value. Basically, I'd end up using this type for ALL the keys in that colors file.

cchaos replied:
To clarify that first question would doing:
`textPrimary: ColorModeSwitch | string;` strongly type to ensure that:

`textPrimary { LIGHT: string; DARK: string }`

thompsongl replied:
`ColorModeSwitch` is a generic type that assumes `string` to start, and results in `{ LIGHT: string; DARK: string }`
Or you can do `ColorModeSwitch<number>` to get `{ LIGHT: number; DARK: number }`.
So  `ColorModeSwitch | string` results in `{ LIGHT: string; DARK: string } | string`.

Should be possible to do this automatically, but I'm not sure yet if the more complex examples work, like `ColorModeSwitch<_EuiThemeBaseColors>`. Because the result would be something like
```
{ 
LIGHT: {
  primary: string;
  accent: string;
  success: string;
  ...
}; 
DARK: {
  primary: string;
  accent: string;
  success: string;
  ... 
}} | { 
  primary: string;
  accent: string;
  success: string;
  ...
}

thompsongl committed:
strict; matching theme shapes

thompsongl commented:
Added `StrictColorModeSwitch` for cases where we don't want the automatic union.
Updated Amsterdam to match the default shape.

This is ready to merge in your main branch if you're good with the general approach

cchaos reviewed (approved):
💯
